src/newgrf_industries.cpp

```diff
@@ -92,10 +92,11 @@ uint32_t GetIndustryIDAtOffset(TileIndex tile, const Industry *i, uint32_t cur_g
 static uint32_t GetClosestIndustry(TileIndex tile, IndustryType type, const Industry *current)
 {
 	uint32_t best_dist = UINT32_MAX;
-	for (const Industry *i : Industry::Iterate()) {
-		if (i->type != type || i == current) continue;
 
-		best_dist = std::min(best_dist, DistanceManhattan(tile, i->location.tile));
+	for (const IndustryID &industry : Industry::industries[type]) {
+		if (industry == current->index) continue;
+
+		best_dist = std::min(best_dist, DistanceManhattan(tile, Industry::Get(industry)->location.tile));
```

Comment:
```suggestion
	if (type >= NUM_INDUSTRYTYPES) return best_dist;

	for (const IndustryID &industry : Industry::industries[type]) {
		if (industry == current->index) continue;

		best_dist = std::min(best_dist, DistanceManhattan(tile, Industry::Get(industry)->location.tile));
```

`type` can sometimes be passed as 240 or 255.

Reply:
Indeed `type` can be `IT_INVALID` with newGRFs.